### DISCUSSION TITLE
Bad optics in my fisheye camera, how to fix images?

### DISCUSSION THREAD

**Discussion by jumpjack** [ImageMagick/ImageMagick#4076] [Digital Image Processing]
I think optics of my "360 camera" (235°FOV) are bad, I see an "unexpected distortion" in them.
Look at what happens when shooting at a checkboard:

![image](https://user-images.githubusercontent.com/1620953/129471042-1c65bd18-00b2-4f5b-a645-24eb7a982d3b.png)

I would expect squares to get bigger and bigger as they get closer to the lens:
![image](https://user-images.githubusercontent.com/1620953/129471123-cd3b4105-3fb4-4959-8686-baa9f0d2727f.png)

Instead they suddenly start shrinking.

Can I fix this (possible) defect using IM?

**Comment by snibgo** [comment]
What version of IM, on what platform?

Yes, IM can do any distortion (or, if you prefer, "undistortion", or "projection") you want.

A conventional camera projects an image to a sensor which is a plane perpendicular to the lens axis. An object that is at 90 degrees from the lens axis would project to an image point on the sensor that is infinitely far from the centre of the sensor.

Your camera might have multiple plane sensors. Whatever sensors it has, the resulting image seems to be projected to a hemisphere, and from there with rays parallel to the lens axis to a plane perpendicular to the lens axis, with some kind of tweaking.

This doesn't mean the projection is wrong, merely unexpected. 

What camera do you have? What does exiftool say about a file created by the camera?

An internet search might find the camera's projection, and suggest possible alternatives.

**Comment by jumpjack** [comment]
I use  IM7 on windows,  but my camera is a Sunchip/allwinnerV3 clone with no optics data in exif, so I have to manually determine its peculiar distortion.
But it's a very cheap camera (60$), it can't have a very advanced 3d sensor.

**Comment by jumpjack** [comment]
>An object that is at 90 degrees from the lens axis would project to an image point on the sensor that is infinitely far from the centre of the sensor

super fisheye lens (>180):
https://petapixel.com/assets/uploads/2019/02/Screen-Shot-2019-02-07-at-11.32.03-AM-1.jpg

**Comment by fmw42** [comment]
See https://360rumors.com/sunchip-panorama-xdv360-cheapest-4k-360/ and https://www.youtube.com/watch?v=vsqJd_5a4aE

**Reply by jumpjack** [reply]
more or less. Mine is more like this one:
https://m.media-amazon.com/images/I/71usS-MmzML._AC_SL1500_.jpg

**Comment by snibgo** [comment]
From those links I suppose there is a single sensor with a fisheye lens that "sees" more than 180 degrees. I can't see anything that describes the projection, which could be stereographic or equidistant or equisolid angle.

Can you post a sample image, such as the "chequerboard" image, direct from the camera without the annotation?

I won't be able to help for a day or so.

**Comment by jumpjack** [comment]
![image](https://user-images.githubusercontent.com/1620953/129525892-6ecc040d-445f-4f98-a5e0-9fe4f25608a8.png)

Did some measurements:

![image](https://user-images.githubusercontent.com/1620953/129525833-34221c0f-6db5-4bc4-8f30-d91469ef472a.png)

**Reply by jumpjack** [reply]
Can this page be of any help? http://www.pierretoscani.com/fisheyes.html

**Comment by snibgo** [comment]
Rows in the chequerboard become closer as they are further from the camera, because of perspective. Things that are further away from the camera look smaller.

So from a roughly central point, near the "4", the distances decrease as we move toward the centre of the image.

Your lens covers more than 180 degrees. It sees slightly behind itself. So the points near the edge of the image are also further away than the "4". So, for the same reason of perspective, those look smaller. Suppose the camera could see all the way behind itself, then the distant rows of the chequerboard behind the camera would be very close together.

So you want to distort the image so that the perspective of the image behind the camera becomes false, with a scale that increases as we go further behind. You want the distances to increase, even though by the rules of perspective they should decrease.

There are many ways of doing this. The following code "unrolls" the image ("-distort depolar"), and creates an absolute displacement map where the green channel (which controls vertical displacement) is a gradient that is distorted when we are more than halfway from the image centre. The distortion is a power function. The unrolled image is distorted by the map, and rolled back up ("-distort polar").

Windows BAT syntax:
```
magick ^
  129525892-6ecc040d-445f-4f98-a5e0-9fe4f25608a8.png ^
  -distort depolar -1,0 ^
  -set option:WW %%w ^
  -set option:HH %%h ^
  ( -size %%[WW]x%%[HH] gradient:Black-White -rotate -90 ^
    -channel RGB -separate ^
    -delete 1 ^
    ( -size %%[HH]x1 gradient:Black-White ^
      -fx "u<0.5?u:u-pow(u-0.5,1.5)" ^
      -rotate 90 -scale "%%[WW]x%%[HH]^!" ^
    ) ^
    -insert 1 ^
    -combine ^
  ) ^
  -compose Distort -composite ^
  -distort polar -1,0 ^
  out.jpg
```
![out](https://user-images.githubusercontent.com/33812368/129645403-28cc42c6-c8b5-411f-ba3a-8046620a0288.jpg)

**Reply by jumpjack** [reply]
can you post the uncropped image? Lines 1 and 2 disappeared.
But I don't know if this solve my final problem: upon reprojecting my original image to equirectangular and viewing it inside a VR viewer, objects close to the border appear squeezed to half their height.
I thought the  fisheye-equirectangular filter of ffmpeg which I used was too generic for my specific camera, then I saw how much squeezed the objects appear already in the fisheye image, so I started to think that the problem is in the source fisheye. 
And I still think so: if your theory was right,  line 3 should be as high as line 5, line 2 as line 6, and line 1 as line 7.

**Comment by snibgo** [comment]
> can you post the uncropped image? Lines 1 and 2 disappeared.

I didn't crop the image. Rows 1 and 2 are visible in the bottom corners. I enlarged pixels that were more than halfway from the image centre. This pushes pixels outwards, so many go beyond the original boundaries. To automatically resize so the output contains the complete input, change the final "-distort polar -1,0" to "+distort polar -1,0".

> ... upon reprojecting my original image to equirectangular and viewing it inside a VR viewer ...

Did you tell the viewer the FOV of your camera? Perhaps the viewer assumed the camera is 180 degrees only.

> And I still think so: if your theory was right, line 3 should be as high as line 5, line 2 as line 6, and line 1 as line 7.

My theory is that the camera was positioned slightly above the "4", about a quarter of the way up the dark patch above the "4", in row 5, at the exact centre of this image (x1.jpg):
![x1](https://user-images.githubusercontent.com/33812368/129904647-cb42f163-e880-4e15-9871-e1f5f006031e.jpg)

If this is correct, row 5 should be the widest, and row 4 should be slightly wider than row 6. These are both true.

My theory about your camera may be wrong. I have almost no experience of such cameras.

**Reply by jumpjack** [reply]
> > can you post the uncropped image? Lines 1 and 2 disappeared.
> 
> I didn't crop the image. Rows 1 and 2 are visible in the bottom corners. I enlarged pixels that were more than halfway from the image centre. This pushes pixels outwards, so many go beyond the original boundaries. To automatically resize so the output contains the complete input, change the final "-distort polar -1,0" to "+distort polar -1,0".

Ok, but the resulting image is far from what I need... I'd need a circular image which distorts the world "correctly"...
> 
> > ... upon reprojecting my original image to equirectangular and viewing it inside a VR viewer ...
> 
> Did you tell the viewer the FOV of your camera? Perhaps the viewer assumed the camera is 180 degrees only.

I think it is not possibile, but I reprojected the fisheye to equi using 235 as FOV, but it's not enough, and now that you showed me the reprojected checkboard, I get why: do you see how the "rectanguli" of the checkerboard are not "equi"? (i.e. rectangles are not identical... and they should not even be rectangles!) as instead they are in a standard equirectangular grid?
![image](https://user-images.githubusercontent.com/1620953/129942511-d2d7418f-825d-4f07-9849-e2077a1ee009.png)

> 
> > And I still think so: if your theory was right, line 3 should be as high as line 5, line 2 as line 6, and line 1 as line 7.
> 
> My theory is that the camera was positioned slightly above the "4", about a quarter of the way up the dark patch above the "4", in row 5, at the exact centre of this image (x1.jpg):
> ![x1](https://user-images.githubusercontent.com/33812368/129904647-cb42f163-e880-4e15-9871-e1f5f006031e.jpg)
> 
> If this is correct, row 5 should be the widest, and row 4 should be slightly wider than row 6. These are both true.
> 

You don't need to guess, I measured everything, you can see the measurements (in cm) in the image above; it was just a qualitative image; using the measures I got this quantitative image:

![checkerboard-reference](https://user-images.githubusercontent.com/1620953/129953440-b6bac7ef-0fd0-456d-8cce-0f0ade7eac38.png)

It results that the top of the camera was slightly above edge between lines 2 and 3. It's confirmed by the equirectangular version: a perfectly straight horizontal line would pass right above this edge.

I also think now that we should focus on the equirectangular version rather than on the fisheye, as it should be "simpler" to properly distort in cartesian coordinates rather than circular coordinates.

I compared the equirectangular checkerboard to a standard equirectangular grid, and the distortion of my camera is now fully clear. I took as reference cell 4 only because it happened that its bottom edge was in line with cell 4 in reference grid, once I placed the red line just above line 2.


![checkerboard-equirect](https://user-images.githubusercontent.com/1620953/129953272-c81d46ad-5e44-40c2-a727-3e1b2753c362.png)

Question is: can this reference image be used to write  some commandline which corrects the checkerboard image in such a way that :
- all squares in central column are squares and not rectangles
- all squares in central column have same height
- all columns have same width


I also noticed another thing:

![checkerboard-bent](https://user-images.githubusercontent.com/1620953/129953260-f7ee0a4a-9443-483a-918a-3a87968cc417.png)

**Comment by jumpjack** [comment]
I built this "distortion curve"... but I have no idea of what to do with it.

![checkerboard-calib](https://user-images.githubusercontent.com/1620953/129956999-07df4f85-cc87-4f3c-8ca7-273d4d6091ea.png)

**Comment by fmw42** [comment]
See the forward fisheye mathematics at http://blog.yeezia.com/?post=91

**Comment by snibgo** [comment]
We are doing too much guessing from a single image.

From your diagram, you seem to assume the nodal centre of the lens it at the front of the lens. But is it? Have you tested this? To test: take a photo with a near object exactly aligned with a distant object. Rotate the camera about a vertical axis through the assumed nodal centre. Take another photo. The objects should align in both photos. Repeat for other possible nodal points; the objects should not align in those photos.

The lens may not be exactly one of the usual types (eg stereographic or equidistant or equisolid angle). To test: take photos of a regular grid that is perpendicular to the lens axis, at different distances. (Example grids: brick walls, office blocks.) With different assumptions of the fisheye type, try to convert to rectilinear projection. Which fisheye type is most correct, with what FOV (angular Field Of View)?

Rotate the camera so the lens axis is parallel to the grid wall. Mark the spot on the wall that is at the camera position. Take photos. This will tell you how far the camera can see behind itself, and what distortions occur there.

The lens may not be axially symmetrical. To test: take a photo, rotate about the lens axis, take another photo. The photos should be identical, within the image circle.

**Comment by snibgo** [comment]
> Question is: can this reference image be used to write some commandline which corrects the checkerboard image in such a way that :
>    all squares in central column are squares and not rectangles
>    all squares in central column have same height
>    all columns have same width

Yes, but a script, not a simple command line. The result will not be a circular image.

For example here (x2-4.jpg) is a version where "all squares in central column have same height", though they are not squares. We could then make "all columns have same width". We could then apply a vertical distortion to make the top and bottom lines straight and horizontal, which would make the squares in the central column into squares. It might make all the other squares into squares.
![x2-4](https://user-images.githubusercontent.com/33812368/129989958-ff05b80a-1e06-4172-a2d5-f972c0724b33.jpg)

**Comment by jumpjack** [comment]
Last image looks good about central column... but not about"something else" which I don't know, and I am out of ideas...
If you load an image in this page you can see how I see in a VR viewer (but it must have 2:1 ratio):
https://photo-sphere-viewer.js.org/guide/cropped-panorama.html#playground

**Comment by snibgo** [comment]
In my previous post, I showed that the image can be distorted so each chequerboard row has the same distance. This is a vertical distortion of the depolar ("unrolled") image, or a radial distortion of the original fisheye image, out2-4.jpg:
![out2-4](https://user-images.githubusercontent.com/33812368/130102582-ed4b6d79-dbce-4926-92d9-6f21ce0589e6.jpg)

The distortion is done with an absolute distortion map. You can use the same distortion map for any 819x819 image from the same camera, to get the same radial smoothing effect.

The other distortions would be specific to this image, and not applicable to other images from the same camera.

**Comment by jumpjack** [comment]
After studying a bit some equirectangular grids, I am no more sure of how the image should appear (both in fisheye or equi projection), that's why i posted the site where to check by yourself is the results you get are good, because I have no more ideas about which transformation I actually need to make the original fisheye image properly visible in a viewer.
